### ISSUE TITLE
bad experience for search

### ISSUE THREAD

**Issue by alvenchen** [Microsoft/vscode#20530]
- VSCode Version: Code 1.9.1 (f9d0c687ff2ea7aabd85fb9a43129117c0ecf519, 2017-02-08T23:31:51.320Z)
- OS Version: Darwin x64 15.6.0
- Extensions:

|Extension|Author|Version|
|---|---|---|
|cpptools|ms-vscode|0.10.1|

---

Steps to Reproduce:

1.when i use global search, it's not easy to type whole file path and name in input box,
  hope that there is a convenient way to  use highlight in search current file by typing "<current file>"

**Comment by roblourens** [comment]
Sorry @alvenchen, I don't quite understand what you're asking for

**Comment by alvenchen** [comment]
@roblourens    Thanks for reply.I wonder is there any convenient way to  use global search when searching "current file" or "all open file".

**Comment by roblourens** [comment]
Expanding this to cover searching open files or searching git modified files, or some other class of files.

**Comment by yuval-gilboa** [comment]
I would also request that you define separate command for starting a search in each search scope, so i can map separate keyboard shortcuts for searching in the different scopes (current file, current folder, open files, all files, etc.). If we have that, it would suffice to cover the use case i requested in #40523, which you've closed.

**Comment by yuval-gilboa** [comment]
Any word on this? I don't see much progress for this feature request.

To clarify, we're looking for something like this:

![image](https://user-images.githubusercontent.com/4710322/35268433-d05d8cca-0031-11e8-8def-5a421d24caf6.png)

Kind of dropdown menu that allows you to select the search scope in the search panel. Possible search scopes can be:

![image](https://user-images.githubusercontent.com/4710322/35268508-fbd05b80-0031-11e8-92b9-163badf9b069.png)

In addition, expose a command for each search scope to allow initiating a search in that scope quickly through keyboard shortcut.

**Comment by gitnix** [comment]
@yuval-gilboa That would be great. In addition to the fields you have listed, I would imagine custom search scopes (where user defines exclude/include files) also being available. "tests", for example, would be a good custom scope. I have a separate issue opened for custom search scopes #42350.

**Comment by ckelley7** [comment]
This is necessary as I use it all the time in VS.NET

**Comment by malfusion** [comment]
@roblourens Do you think limiting the scope of the "files to include" field, to the current files/files would be enough to accomplish this? I already have a version of VSCode where i have modified it to do this. I can submit a PR soon if you think that's enough.

**Comment by atif089** [comment]
@malfusion I like the idea, but wouldn't that discard all the unsaved changes in the editor?

**Comment by malfusion** [comment]
@roblourens so the "files to include" is a glob pattern that is searched. So the trick here would be to find the files that are required, generate the glob pattern and then update the "files to include" field with it. And the search would only search the matching files.

@atif089 if the question meant "would the search disregard the current file's unsaved changes", then I'm not sure yet. I'll have to check. But to my knowledge, the search is performed directly on the file system, so only saved changes might be searched.

**Comment by bkarlson** [comment]
was amazed (in bad sense of this word) to discover VSCode cannot do it... it's 2018 after all, seems to be a common sense function present in any other code editor or IDE..  bumping issue to attract attention to it.

**Comment by damienleroux** [comment]
+1

**Comment by e105-gamma** [comment]
+1

**Comment by orecht** [comment]
I used "search in open files"  a lot in Notepad++. It is especially useful for searching multiple csv files (or other data files).

**Comment by rdewolff** [comment]
Love this feature!

**Comment by shahidmau** [comment]
+1 must have feature.

**Comment by Instagit** [comment]
How about changing the scope of "files to include" with key words like ````@open````, ````@closed````, ````@modified````, ````@unsaved````?

It wouldn't require any UI changes and is easily extendable, so if in the future you decide to make files searchable by modify date, that could be ````@modified(2d)```` or ````@modified(48h)```` for files that have been modified within the last two days.

I can see this being a very useful advanced feature without needlessly cluttering the UI.

**Comment by BillDenton** [comment]
From #18796 : 
"Request ability to search only in the files that are loaded in the editor. In addition it would be good to have a search results window for find (in current file Ctrl+F) so that you can see a list of matches with some context and the ability to copy the search results and click on them to go to the match. This should also work with replace. May be there should be another window like Output, Problems, etc. Ideally this would be able store multiple sets of search results in multiple windows. See SlickEdit for example."
I think the ideas have cover most of #18796, but possibly not suggestion of multiple search result windows. It very useful to have multiple search results, particularly when working with log files.

**Comment by roblourens** [comment]
I think https://github.com/Microsoft/vscode/issues/16488 might cover the rest of that, keeping track of multiple independent searches

**Comment by mromanch** [comment]
I really love the suggestion from @Instagit!!!

I find it superior to just a UI dropdown, because I can type something like `@modified, @open, src/**/*.xml` into either the **files to include** or the **files to exclude** and anything else that takes file filters could accept the same patterns.

I also love how the idea includes the ability of extensions that extend the filter functionality.  (_Dreaming_ :sleeping::heart_eyes::joy::stuck_out_tongue_winking_eye: _of GitLens support for `@git(52b15e9)` allowing me to search within a specific commit._)

Not to mention, if you can accept pre-defined filters, than it's not hard for me to imagine user-defined filters, e.g.:

```
"search.filter": [
    { "config": ["**/*.config", "src/**/*.cscfg"] },
    { "deploy": ["**/deploy", "@config" },
    { "test": ["**/test", "!**/test/log", "!**/test/out"] },
    { "modified": ["@staged", "@unstaged", "@untracked"] },
    { "unmodified": ["!@modified"] },
    { "src": ["!@deploy", "!@test", "@modified"] },
    { "prod": ["@src", "@deploy", "!@modified"] }
]
```

Some things to note about the above suggested syntax:
- filters are an array of patterns
- filters can be defined from filters
- filters support _gitignore_ syntax including negation (i.e. same as existing filter fields)
- filter pattern order matters , e.g.:
  - the `@src` filter includes `@modified` files that might be excluded by `!@deploy` or `!@test`
  - the `@prod` filter includes `@deploy` back in, but removes `@modifed`

Given the above syntax, the builtin filters only need to support some basic operations like `@open` and `@unsaved`.  Then `@closed` simply becomes `!@open`.

I'm not sure if `@folder`, `@workspace` and `@project` make sense.  In the find panel, anything like `@folder`, or even a UI dropdown will lack context.  However, a keybinding for something like `find(@folder)` probably does make sense.

I feel like minimum Git support is important.  Filters for `@staged`, `@unstaged`, `@untracked` and default definitions for `"modified": ["@staged", "@unstaged", "@untracked"]` and `"unmodified": ["!@modified"]` would be really nice.  I don't know if these are things that **must happen** in this feature, or just things that **must be enabled** and support gets left to an extension.

**Comment by roblourens** [comment]
The custom filters are a really interesting idea.

**Comment by imnickvaughn** [comment]
+1

**Comment by haugerbr** [comment]
I specifically would like to see at least a **current file** and **all open files** option. It doesn't seem like this would be hard to implement. What files would I be looking at in vscode if I wanted to implement this and submit a pull request?

**Comment by ivanp2015** [comment]
+1 for this feature. Especially would like to have "all open files" and "modified files"

**Comment by ivanp2015** [comment]
Any plans to implement this in 2019?

**Comment by dustiniapp** [comment]
+1

**Comment by itsrachelle** [comment]
+1 all open files search option would be great!

**Comment by deepaktatineni** [comment]
Any plans to includes this feature?

**Comment by edespong** [comment]
I will be taking a stab at this based on the ideas of @Instagit and @mromanch

**Comment by roblourens** [comment]
Sure. Just please discuss your approach with me before spending a lot of time on it.

**Comment by edespong** [comment]
**Basic approach**
The approach I have looked at is to extend the `QueryBuilder` to make use of a new 'token expansion' class. The current builder splits the query into glob patterns and makes IExpressions out of them. The idea would be do the expansion between the split and conversion to IExpression. For example, the include pattern `*.ts, @open, *.css` would leave the `*.ts` and `*.css` intact while expanding `@open` into the file paths of the currently open files (though issue https://github.com/Microsoft/vscode/issues/48958 is blocking for this to work).

Tokens that simply expand into normal globs can also be supported via config.

**Extension**
The actual token expansion will not have any hard-coded tokens, but these would rather be supplied by config and extensions. The ones I have looked at are the editor and git, to support for example open and unsaved files (editor) and changed, staged, etc (git). The extensions esentially use the command framework and the token expander use the ICommandService to call the extension and get  the expanded token back.

**Syntax**
The default @ syntax for vscode seems to be `@token:parmater`, e.g. `@editor:unsaved`. This can then be shortened or aliased for the most common use-cases, see Configuration.

One question would be how to handle use-cases where multiple parameters are needed and what the delimiter should be. E.g. searching for files changed within the last day. Like `@file:modified;24h` or `@file(modified, 24h)`.

**Configuration**
Pretty much follows @mromanch idea. Can either define hard-coded globs, combinations of previously defined configuration or reference tokens defined by extensions. Example:
```
{
    '@config': ['**/*.config', 'src/**/*.cscfg'],
    '@deploy': ['**/deploy', '@config'],
    '@modified': ['@git:changes, @git:staged, @git:untracked']
}
```

**UI**
Right now, there is only 'files to include', but other programs like Visual Studio also has a 'scope' (e.g. Entire Solution, Open Files, etc). Does it make sense to differentiate between them or will it only be confusing?

***Auto-completion?***
Have not looked into this.

**Comment by roblourens** [comment]
This sounds great, however it sounds like https://github.com/Microsoft/vscode/issues/48958 will be blocking most of this. Unfortunately that's harder to fix than it sounds because it will involve changing extension API and I will need to drive that.

**Comment by edespong** [comment]
Yes, I do not see any good way of solving this issue without having https://github.com/Microsoft/vscode/issues/48958 in place. I will keep an eye on if/when it is fixed.

**Comment by adjenks** [comment]
https://stackoverflow.com/questions/42011289/how-to-search-only-in-currently-open-files-in-vscode
https://stackoverflow.com/questions/56086094/how-do-you-search-within-all-open-files-in-visual-studio-code-not-in-a-folder

**Comment by eminozlem** [comment]
bump . i really need this, this could be the only reason if I ever go back to Dreamweaver. "Search in open files" is very handy 😄

**Comment by abhay-gooru** [comment]
that is really a blocker, i came from sublime to vscode, before that notpad ++, vs.net etc all have them, but vscode does not. I have already going back to sublime or searching another alternatives. Soon i find that will dump VS code

**Comment by Bagnall** [comment]
Support request for "search in open tabs". Please!

**Comment by SuperV1234** [comment]
+1 for search (and replace) in open files, it is extremely needed.

**Comment by gino8080** [comment]
bump! +1

**Comment by FabianLauer** [comment]
Guys, please stop "bumping". It triggers notifications for everyone involved in the thread and doesn't add any value. Just hit 👍 in the top comment instead.

**Comment by carehart** [comment]
One problem with hitting the thumbs up in the top comment is that that top comment (the issue title itself) doesn't convey what nearly all the rest of the comments are getting at: a desire to have a feature to "find in open files/tabs". The OP worded the issue title and even their initial post in a way that does not convey this broader problem. I wish we could more clearly be voting for this broader problem/feature request.

Is anyone (who is a party to this thread) aware of any other issue or PR where attention to this broader issue (a desire to find against open files/tabs) maybe has moved? so that we can drive our support there, more clearly?

**Comment by roblourens** [comment]
@carehart I consider this to cover that request, and I added a note to the OP to clarify that.

**Comment by carehart** [comment]
Great. Thanks.

 

/charlie

**Comment by stderr-to-devnull** [comment]
So this has been open for 2 1/2 years, is it so monumentally difficult to implement? Other more basic editor already have this (Notepad++ for example). I need to be able to search in all open file tabs. How many thumb ups are necessary? Maybe I can talk with my family and friends to create accounts and support this request.

**Comment by eminozlem** [comment]
Yea I am looking forward to this. Every time this thread gets replied I get
excited, only to find out it was yet another bump

stderr-to-devnull <notifications@github.com>, 25 Eyl 2019 Çar, 14:34
tarihinde şunu yazdı:

**Comment by edespong** [comment]
@roblourens If it is necessary to support an inclusion pattern like specifying both a glob and specific files, would that even be possible today with ripgrep?  

If that is not needed, could we not "just" extend getRgArgs to work with absolute paths and disallow the combination of absolute paths and globs?

**Comment by aetonsi** [comment]
Hi,
i found out that i can "search only in open files" by using a backslash instead of a forward slash.
Is it a feature? Is it a bug? (if it's a bug i'll open an issue, obv)
Demonstration screenshot attached (i was looking for php files inside the "backend/app" folder)
> VSCode-win32-x64-1.38.1, extensions disabled, exclude settings/ignore files disabled

![image](https://user-images.githubusercontent.com/18366087/65981344-2be34080-e479-11e9-9916-d5e780995d03.png)

**Comment by francoismassart** [comment]
Hi @aetonsi, I just tested this technique on VS Code on macOS and it does NOT work 👎 

Here is my config...

Version : 1.38.1
Validation : b37e54c98e1a74ba89e03073e5a3761284e3ffb0
Date : 2019-09-11T13:31:32.854Z
Électron : 4.2.10
Chrome : 69.0.3497.128
Node.js : 10.11.0
V8 : 6.9.427.31-electron.0
OS : Darwin x64 16.7.0

Cheers

**Comment by robert1826** [comment]
I've found a simple solution but it might not work for all scenarios.

scenario: search for a keyword in modified files between two branches (like master and current branch) - I'm using gitlens
steps:

1. compare the two branches in gitlens to find the modified files in the compare section
2. right click on 'x files changed' and click open files
3. to to explorer in vscode
4. select the open all files under open editors section
5. right click on the selected files and click copy path (or relative path)
6. open an new untitled file and paste the paths and then proceed to join all the paths by commas
7. open search all files and paste the comma joined paths in the files to be searched section

this idea could somehow extend on multiple levels but as you can see it's not the best and I'm definitely waiting for an official solution from the vscode team

**Comment by jesseko** [comment]
Sublime has a nice model for this that could work for VSCode too.  

The text input for which files to search in is paired with a "..." control to get access to these special options.  When one is selected, they insert a special entry into the text field.  This lets these be mixed with other custom search patterns and makes it easy to always see the full set of search patterns in one place.

Dropdown at right
![image](https://user-images.githubusercontent.com/153062/68340549-7de03d00-009b-11ea-91bb-fce85b9d30f3.png)

Choose "Add Open Files"
![image](https://user-images.githubusercontent.com/153062/68340614-9cdecf00-009b-11ea-9c6b-0209a1c5ab44.png)

&lt;open files&gt; is added to the text field
![image](https://user-images.githubusercontent.com/153062/68340634-a9fbbe00-009b-11ea-806f-6f432214271d.png)

You can have multiple of these together, and it's easy to delete them at any time
![image](https://user-images.githubusercontent.com/153062/68340989-550c7780-009c-11ea-9c64-23a9fa045607.png)

**Comment by edespong** [comment]
I have created two branches on my fork, the first one [Support for tokens expanding to search patterns ](https://github.com/edespong/vscode/commit/180056533ddd97f95cfd953751a3505ea697e4c6) contains basic support for @mromanch idea of pre-defined filters. This branch is independent of the blocking API issue, and if @roblourens thinks it is a good idea, I could create a new issue to which a PR can be submitted. 

The second one [(Experimental) support for targeting sets of files](https://github.com/edespong/vscode/commit/ef3d7ccd9c8c0f35b463b4b45fa2185f27023e47) builds on the first one, but adds support for editor (open, unsaved files etc.) and git (modified, untracked, etc). The code contains a very ugly workaround to the #48958 issue.

**Comment by gjsjohnmurray** [comment]
Maybe this could leverage the new [Search Editor](https://github.com/microsoft/vscode/issues/23931#issuecomment-559227491) that landed as a preview in 1.41. Then we could have a way of re-searching (with a different search string) in the exact set of files that an initial search found its matches in.

cc @JacksonKearl

**Comment by JacksonKearl** [comment]
@gjsjohnmurray for the time being, an extension could pretty easily contribute such a command. I can fill in with more details if anyone is interested.

**Comment by DBjelovuk** [comment]
Would love a way to limit search to only changed/modified files.

**Comment by electrovir** [comment]
@JacksonKearl do you know of one that already does this?

**Comment by orecht** [comment]
If mean one text editor that does this, Notepad++ does that. The Find box
has a button "Find Al in Opened Documents"

**Comment by JacksonKearl** [comment]
@electrovir No I don't know of any extensions building off of the Search Editor. This is probably a good thing for now, as it is still experimental and subject to change.

**Comment by dnecklesportfolio** [comment]
This would be so helpful, im on a project right now and I am surprised this hasnt been added with the lengthy discussion had so far.

**Comment by dnecklesportfolio** [comment]
workaround, i search in every file manually.. which is tedious but **once you do it once**, go to each file and search.. each file comes up in the results on the right.. and then you have the feature we've been desiring

**Comment by eminozlem** [comment]
yea good workaround let me just open 56 files and search every single one
of them XD

Dwayne Neckles <notifications@github.com>, 22 Oca 2020 Çar, 23:26 tarihinde
şunu yazdı:

**Comment by dnecklesportfolio** [comment]
Ha Good point.. I meant it workaround for if you have a few files under 10.. Hope Vscode considers this.

**Comment by Sadi58** [comment]
Another workaround 😉
https://notepadqq.com

**Comment by DavidMulder0** [comment]
Btw, for those who want to search in their (staged git) changes, you can use

    git diff-index -U --cached -G <regexp> HEAD

Sadly as it's a diff command technically, it will include the `a` and `b` in the paths, so ctrl+click won't work from the integrated terminal 😢

**Comment by nemile** [comment]
> Btw, for those who want to search in their (staged git) changes, you can use
> 
> ```
> git diff-index -U --cached -G <regexp> HEAD
> ```
> 
> Sadly as it's a diff command technically, it will include the `a` and `b` in the paths, so ctrl+click won't work from the integrated terminal 😢

To search in git modified/staged files only, I copy the `git status` file names in the result (all at once), then paste it in the editor and manipulate all lines at once to get a unique line with all file names joined by a comma*.
I then copy this line in the "files to include" of VScode, and thus can search in only the modified/staged subset of files.

*On mac OS, Cmd+A to select all, then Options+I to get a cursor at the end of each line. Then just add a comma on each and Del. You get one line with all file names.

**Comment by pgarciacamou** [comment]
@nemile 

👉 `git diff --name-only` 👈 

alternatively: `git diff --name-only > temp.txt` 

Still not ideal.

**Comment by VishmayShah** [comment]
"file to include" may actually work if you can enhance the search and context menu on the page.
Somewhat like this 
![Screenshot_17](https://user-images.githubusercontent.com/13658837/74008161-32913f80-49a6-11ea-8d41-a262d1acb344.png)

Where 
1. multiple files can be included. 
2. The full path is used, so other pages having the same file name doesn't get affected(child, sibling or parent)
![Screenshot_20](https://user-images.githubusercontent.com/13658837/74008429-cd8a1980-49a6-11ea-8003-385cb13fe0fe.png)

Share your thoughts

**Comment by krishna81m** [comment]
Two years and no update or a plugin, I end up using a work around to use specific folder for all of my notes and only search that folder in a dedicated window to limit search results

**Comment by JacksonKearl** [comment]
@krishna81m you can search in a specific folder by putting it in the “files to include” box. This is about targeting files that can’t be classified using globs.

**Comment by dnecklesportfolio** [comment]
Dont feel too bad vscode users, sublime, atom, and brackets do not appear to do this either.

**Comment by gesslar** [comment]
> Dont feel too bad vscode users, sublime, atom, and brackets do not appear to do this either.

I switched from UltraEdit to this. It's the only feature I'm still missing.

**Comment by evenfrost** [comment]
@dnecklesportfolio FWIW, Sublime had this feature from long ago. You can search in e.g. currently opened editors with `<open files>` placeholder.

**Comment by krishna81m** [comment]
@JacksonKearl , apologies for not being clear, I only want to search in files "that are open" and there isn't a way to do this.

**Comment by VishmayShah** [comment]
My current workaround for this functionality is notepad++. Still looking forward to this.

**Comment by kotpal** [comment]
Will this feature ever be considered for VSCode?

**Comment by JacksonKearl** [comment]
I see the appeal to having something like sublime's `<open files>`. Still not clear what other custom sets would be allowed. Ideas: tracked, staged, modified, etc. But those would need to also work across scm providers, so there are some things to consider about how that works.

Also not sure what the UX would be for inputting them. We have custom flags in the extension search and settings search identified by `@`, but those searches are both somewhat constrained in terms of what you might search for. In the workspace search however you should still be able to search in the literal file `<open files>.txt` or `foo/@open`, etc.

**Comment by kotpal** [comment]
Let's start with 'Search in Open Files' at the very least. That seems to be the most requested feature. 

We can't work on all the features - like integration with scm, searching different artifacts and such - right off the back.

**Comment by JacksonKearl** [comment]
@roblourens any thoughts on how a syntax for these might look? If its opt-in I think `@open`/etc. would work well, but we'd also need some escape mechanism. (`@@`?)

**Comment by roblourens** [comment]
There is already an open PR that partly implements this with @ but it's blocked on needing fixes in search API land

**Comment by VipSaran** [comment]
Cmon, it's ridiculous. The search looks in open files by default **only** if the file was focused after it was opened!

We need an option to automatically look in all opened files (in the window).

**Comment by kglickman** [comment]
I would love a recently used list for the "files to include".  I am constantly toggling back and forth between a few different directories.

**Comment by JacksonKearl** [comment]
@kglickman you can use the arrow keys to scroll through your history, does that help?

**Comment by kglickman** [comment]
Yes, it definitely does!  It would be nice if there was some kind of visual indicator that you could do this ...

**Comment by carehart** [comment]
Going back to the original post and the few that followed, which were focused on how to search "only open files", I'll share **something I just discovered**. 

**If you have only opened files in vscode directly,** by way of using File>Open, or Open Recent, or by opening a file from the OS into vscode, etc., THEN a search/find in files WILL indeed search ONLY (and all of) those opened files.

But it's **if you use File>Open Folder**, that's when things change about search (and of course many other ways about the UI). It THEN presumes to search first all files/folders in THAT folder, and you can't tell it to ONLY search open files. To be clear, it DOES ALSO search open files--and it shows the results of that as "other files", below the search of the "opened folder".

I suspect it's those folks (who use that valuable "open folder" feature) who are calling for an option to be able to limit the search to ONLY open files. I didn't see it mentioned in any comments above, so I wanted to offer this distinction.

I share it also for those coming from other editors, who may tend to only work in a few files, perhaps spread over multiple folders, and they may want to by default have it search THOSE (and only those) open files. In that case, just close the "open folder" feature (File>Close Folder), and then the search will indeed be only for files you open otherwise. 

All that said, I can see why folks are STILL waiting for an option to tell it to search ONLY open files, for when they DO have reason to have used the Open Folder feature. 

Hope that's helpful.

**Comment by eminozlem** [comment]
This is never going to happen is it

Charlie Arehart <notifications@github.com>, 15 May 2020 Cum, 20:57
tarihinde şunu yazdı:

**Comment by fadimounir** [comment]
Any updates on this? Would love to have this feature

**Comment by narkoze** [comment]
Waiting for this 😇

**Comment by jockrow** [comment]
could be some like this:
![image](https://user-images.githubusercontent.com/30563443/86124087-26fefc00-baa0-11ea-939c-375d5c135492.png)

**Comment by JacksonKearl** [comment]
@jockrow you can do that already like so:
![image](https://user-images.githubusercontent.com/8586769/86159196-d054ea80-babe-11ea-85b6-049fbbcd917d.png)

**Comment by markdomansky** [comment]
Commenting only to add endorse that we need this feature.

**Comment by eminozlem** [comment]
yea its not happening XD

**Comment by EwSt** [comment]
How is this not a thing? Also why doesn't the "files to include" have a better way to display all files i put it in it?

**Comment by jackmparker** [comment]
Commenting to add support for a "search open files only" feature.

**Comment by jkozera** [comment]
@gitnix you [seemed to have](https://github.com/microsoft/vscode/issues/20530#issuecomment-361739971) the same idea as me :)
> @yuval-gilboa That would be great. In addition to the fields you have listed, I would imagine custom search scopes (where user defines exclude/include files) also being available. "tests", for example, would be a good custom scope. I have a separate issue opened for custom search scopes #42350.

Like here?

![search-scopes](https://user-images.githubusercontent.com/120114/89443382-7fd14c80-d750-11ea-9887-9d5a9893a190.gif)

I'm not sure if such quick PoC hacks are a good idea, but maybe a proof of concept would encourage someone to do a proper implementation. Anyway, there it is:

https://gist.github.com/jkozera/f3bbfdbeb1ffdcc08ef7a02acdfc625b

Note it may be unsafe (it downloads jQuery JS), or may break with different VSCode versions if HTML class names change, for instance. The checkbox also may disappear wen navigating to some other places in the GUI.

Anyway pasting it in devtools (Help -> Toggle Developer Tools) while the exclude input is visible worked for me.

**Comment by machineghost** [comment]
Came here to also beg for this feature, only to discover JacksonKearl just picked it up.  Thank you!  Here's hoping we finally see this super-useful feature in VS Code! 🤞

**Comment by eminozlem** [comment]
I dont know why people suggesting complicated or unrelated solutions. What we need is so simple "search in open(ed) files".   So If  I have "inc/header.php" and "assets/style.css" open in my workspace, I only want to search only in these 2 files. Nothing more, nothing less.

![XyjNz](https://user-images.githubusercontent.com/507708/91241794-68cdbb00-e74e-11ea-97b2-776ae3f5526a.png)

Dreamweaver had this since 12 years. VSCode is fairly new, but has to pick up the pace, soon its gonna be 4 years since this has been requested.

**Comment by edespong** [comment]
> 
> 
> I dont know why ...

It is literally described in the title of the issue and the initiating post

**Comment by machineghost** [comment]
> What we need is so simple "search in open(ed) files".

To borrow a line from the Lone Ranger: "who's this 'we' kemosabe?" ;-)

What I mean is ... when a library/tool author is trying to solve the concerns of many people, I don't think it's the most productive thing to show up and say "hey look at how simple it is to solve *my* problem; ignore all the other problems you're solving and just fix *mine*".

Now at the same time, **I 100% agree that _just_ a "search across open files" would be super userful**! I also 100% agree that the VS Code folks are probably making perfect the enemy of the good here: they are trying to solve a bunch of niche problems at once, instead of addressing a simple and easy to solve problem that would positively impact a huge number of users (and they are using the difficulty of the niche problems to avoid addressing the giant low hanging fruit in the room).

I truly wish that @JacksonKearl would stop dreaming up beautiful and awesome UI that would solve a million tiny problems ... and instead just give all VS Code users one simple checkbox (find across open files) that we'd all *really* appreciate, and that (as you noted) is super common in other editors.

In other words, I think there's really two tickets here: the one you (and I and **TONS** of VS Code users) want fixed ... and a separate one described at the top of the page (that some devs would also love to see fixed).  From my perspective, it would make a lot of sense to fix that first one now, and let the one at the top of the page languish until it has more demand.

But just ... when you write something like:

> What we need 

Maybe recognize that you can't speak for everyone, and that the VS Code devs aren't just thinking about any one user's needs ... they're trying to solve everyone's.

**Comment by oholimoli** [comment]
Please add this basic feature

**Comment by machineghost** [comment]
@JacksonKearl I truly think you're letting "perfect be the enemy of the good" here ... and it's preventing all VS Code users from getting a really valuable feature

We don't *need* the worlds most complex file filter!  I would imagine that everyone who saw your beautiful mock-up thought "that looks awesome" (I certainly did) ... but it's also crazy scope creep.

It seems to me 90% of the people in this thread need a "find across open files" feature.  All this "perfect dream UI for filtering by any type of file" stuff is  is making it impossible to provide that very basic feature.

If we could just make this ticket "add a checkbox to find in open files" ... it's not that I *don't want* your beautiful, and far more powerful UI ... it's just that I don't want to wait for it while it holds up the much more utilitarian feature everyone needs.

**Comment by JacksonKearl** [comment]
@machineghost thanks for the feedback! 

I definitely agree that small, incremental, work towards the "grand vision" is better than an all-or-nothing approach.

That being said, I think it's important for the work to actually be incremental, so perhaps instead of adding a single-purpose checkbox to the UI for opened files (with all the UI discussion that entails...), a set of `@`-filters could be implemented in core (potentially starting out with just `@open`), deferring discussion of API and extensions contributing filters.

**Comment by machineghost** [comment]
If that delays things further, I think you're still letting perfect be the enemy of the good.  If it's possible to get people the functionality they need now AND do it with @ - filters, GREAT!

I'm just trying to point out that you're making things worse for everyone every time you do anything to delay giving them a basic feature VS Code should have had years ago ... even if you do it out of a desire to give them an even better feature.

**Comment by planetwarevince** [comment]
+1 for an option to make search [and replace] act only on "all open files".

**Comment by JacksonKearl** [comment]
Created https://github.com/microsoft/vscode/issues/106790 to track exploring how to implement this for just opened editors at first.

**Comment by ArturoDent** [comment]
It was recently reported in a SO answer that using `./` in the `files to exclude` accomplishes searching only in opened files?  Please test this.

**Comment by penguin020** [comment]
Work Around: 
Entering "./" in "Files to Exclude" forces VSCode to return only open files.

https://stackoverflow.com/questions/42011289/how-to-search-only-in-currently-open-files-in-vscode

**Comment by machineghost** [comment]
So, again, literally all that's needed here is a checkbox in the UI: every last bit of functionality already exists.

The *only* thing preventing millions of VS Code users from being able to search across open files is the fact that they have to come find this issue to learn how.  Adding a simple checkbox to the UI would solve this.

**Comment by roblourens** [comment]
There are hundreds of open feature requests that would be resolved with "just a checkbox", then where does that leave the UI? And @penguin020's workaround I don't think will really work in all cases.

**Comment by JacksonKearl** [comment]
The problem is that Open Editors searching in general only works on editors that have been viewed this session. If you reload and try the workaround you'll see no results. As it stands that's an unfortunate limitation but overall fine, however if we are to ship a feature that is called "search in open editors", it should work with all open editors, not just the ones that have been clicked on since reload.

**Comment by canvaspixels** [comment]
My colleague just said she'd move to vscode if there was a checkbox for `Find in open files`

**Comment by IvanPizhenko** [comment]
> My colleague just said she'd move to vscode if there was a checkbox for `Find in open files`

I'd rather vote for a dropdown "Search in:" with options like "Workspace", "Open files", "Unstaged changes", "Staged changes" (but at least first two)

**Comment by yuyoyuppe** [comment]
If it's possible to remove the limitation mentioned by @JacksonKearl, it would be extremely useful feature.

**Comment by tomaszs** [comment]
My friend told me he has downloaded VSCode and waits for "Find in files" to install it

**Comment by danavangard** [comment]
I was happily debugging/updating dozens of pages of code simultaneously with Notepad++ on my PC laptop for a dozen years now. I just bought an iMac (2009), yet because NP++ does not have an equivalent for Mac, I was advised by a few coders to try Visual Studio Code. VSC appears awesome for STACK coding, alas, much to my utter dismay, there is no simple dropdown menu like the type advised by IvanPizhenko et al that allows to search/edit/update, among other things, **only the files that are open**. I code UI/UX functionality for web software and lack of such a basic tool within VSC is beyond me. This tool must be _built-in_ and made easily accessible, usable without having to delve into JSON scripting/extension development. Hoping VSC devs will implement this asap before I have to install a Windows 7 partition on my iMac so I can happily use Notepad++ again... .

**Comment by danavangard** [comment]
Woah! Just downloaded Sublime Text 3 (ST3) on my Mac, gave it a test drive, and am already coding like breathing heavy ;-) The UI is super efficient (the opened tabs can be navigated, among others, with mouse click or _mouse wheel_),  is not at all cluttered, is super-fast, opens an independent window whenever you edit parameters unlike what VSC should not do (opening any param file _within_ the opened project - just adds to clutter). Besides the convenient plethora of possibilities ST3 offers, there is a _Find and Replace_ dialog that appends itself to the bottom of document (unlike VCS which floats the dialog on top of the opened files) which offers following options (drum roll...):

1. Folder
2. Include Filter
3. Exclude Filter
4. Open Folders
5. Open Files
6. Current File

... with a _live_ _while-you-code_ editable combination of them all!

To gain a wider audience, VSC should present things much simpler as default by letting Users add UI/UX modules to the main architecture _where and when_ they need more, as they mature. I will use VSC for tasks ST3 does not handle, but the no-frills interface ST3 offers allows me to concentrate much more quickly on actual coding than having to wrestle with the navigational maze VSC provides. Hope this helps.

**Comment by naarcini** [comment]
+1 to the request for "Find in currently opened files". I work on Chromium code and it's often impractical/pointless to search for a term in my opened workspace since that contains a lot of files. Having this feature would make some refactoring scenarios a lot easier.

**Comment by sickdyd** [comment]
+1 I think adding this feature would be really helpful. Thank you VSCode devs for your time and efforts.

**Comment by eminozlem** [comment]
I am out of snarky comments, just bumping so that they know we are still waiting :D hopefully in 2021

**Comment by eminozlem** [comment]
I am out of sarcastic comments - been waiting for years. Just bumping so that they know we are still waiting :D 
I hope they release "search in open files" soon at least, and not go overboard with it. Hopefully we get it in 2021

**Comment by ooobii** [comment]
Big 'ol bump! 

This feature is too convenient to leave deserted! Please add this!

**Comment by atif089** [comment]
> I am out of sarcastic comments - been waiting for years. Just bumping so that they know we are still waiting :D
> I hope they release "search in open files" soon at least, and not go overboard with it. Hopefully we get it in 2021

It's been only 3 years you impatient monster.

**Comment by machineghost** [comment]
It's ok guys: when this feature is finally released (in 2035) it will do EVERYTHING for you.  It will write your code for you, wash your butt for your ... EVERYTHING.

All the scope creep ... while we wait forever for a basic feature every other major editor already has ... will totally be worth it ... in 2035.

**Comment by JacksonKearl** [comment]
Starting with today's insiders a checkbox has been added to specify searching in open editors. The includes/excludes filters will then whittle results from those editors.
![recording (6)](https://user-images.githubusercontent.com/8586769/105537850-62188500-5ca7-11eb-8037-4a8d3fe63adc.gif)

Additionally, a command "Git: Open all Changes" has been added to help with searching in all changes, but I'll be keeping this ticket open until to track creating a good flow for searching in particular file groups.

**Comment by xgqfrms** [comment]
#  update for new version solution

```sh
Version: 1.56.2
Commit: 054a9295330880ed74ceaedda236253b4f39a335
Date: 2021-05-12T17:44:30.902Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Darwin x64 20.3.0

```

![image](https://user-images.githubusercontent.com/7291672/121217884-d8afcb00-c8b4-11eb-843a-c7bebf673fa1.png)

![image](https://user-images.githubusercontent.com/7291672/121218072-009f2e80-c8b5-11eb-8c45-de774b0b99a5.png)

**Comment by Louis7777** [comment]
If we could also have a toggle next to Aa to **ignore whitespace**, that would be great. Don't suggest to use regular expressions. You can't search for entire blocks of code that way. It's insane and not practical at all.

**Comment by mmcintyre123** [comment]
> If we could also have a toggle next to Aa to **ignore whitespace**, that would be great. Don't suggest to use regular expressions. You can't search for entire blocks of code that way. It's insane and not practical at all.

Your use case is not clear.  Why would you need to ignore whitespace when searching for text?  It would indeed seem that very simple regex in the search query would satisfy the vast majority of needs.

**Comment by yuval-gilboa** [comment]
@xgqfrms, @JacksonKearl 

This solution only covers one of the use cases we asked for. A single toggle button is not enough, need a dropdown with multiple search scope options as requested at the top of this thread more than 4 years ago...

Also needed are separate commands for starting search in each of these search scopes, so separate keyboard shortcuts can be mapped for searching under the different scopes

**Comment by Louis7777** [comment]
> 
> 
> > If we could also have a toggle next to Aa to **ignore whitespace**, that would be great. Don't suggest to use regular expressions. You can't search for entire blocks of code that way. It's insane and not practical at all.
> 
> Your use case is not clear. Why would you need to ignore whitespace when searching for text? It would indeed seem that very simple regex in the search query would satisfy the vast majority of needs.

![ignore-whitespace](https://user-images.githubusercontent.com/7398634/123514125-e4afd100-d699-11eb-9e84-5c549aba156c.png)

Such an option is not unheard of. I may need to search for a block of code. For reasons. Maybe I want to apply a fix that requires replacing a block of code, and I don't want to replace it line by line.

Let's say a simple if-else block that contains some string variables. How would you do it with regex? Is it really very simple? Is it fast to think of the regex? And if it is that simple, why wouldn't there be a toggle that does exactly that?

If you know how, I'd really like to have an example or a link that shows how to craft a regex to find a specific block of code.

**Comment by mmcintyre123** [comment]
> > > If we could also have a toggle next to Aa to **ignore whitespace**, that would be great. Don't suggest to use regular expressions. You can't search for entire blocks of code that way. It's insane and not practical at all.
> > 
> > 
> > Your use case is not clear. Why would you need to ignore whitespace when searching for text? It would indeed seem that very simple regex in the search query would satisfy the vast majority of needs.
> 
> ![ignore-whitespace](https://user-images.githubusercontent.com/7398634/123514125-e4afd100-d699-11eb-9e84-5c549aba156c.png)
> 
> Such an option is not unheard of. I may need to search for a block of code. For reasons. Maybe I want to apply a fix that requires replacing a block of code, and I don't want to replace it line by line, or make my search too specific to locate it (e.g. type some phrase that seems unique?) when I know that there's somewhere exactly one block like that.
> 
> Let's say a simple if-else block that contains some string variables. How would you do it with regex? Is it really very simple? Is it fast to think of the regex? And if it is that simple, why wouldn't there be a toggle that does exactly that?
> 
> If you know how, I'd really like to have an example or a link that shows how to craft a regex to find a specific block of code.

Sure no problem. So, when you say find a specific block of code, would you be searching for it exactly as written? Or would you be searching based on part of the contents? Please provide a sample block of code and I will write some regex for you.

**Comment by Louis7777** [comment]
>
>
> Sure no problem. So, when you say find a specific block of code, would you be searching for it exactly as written? Or would you be searching based on part of the contents? Please provide a sample block of code and I will write some regex for you.

Exactly as written, but even in the case of a partial search it could still be a block of code instead of a line or phrase.

Let's say for example the code below. I've intentionally screwed some spaces, since that can happen.

```
function whatever( argument ) {

  let myvar=null;

  if ( argument){
    myvar = '';
  }
  else {
    myvar='this is a string that is not unique to this section';
  }

  return myvar;
}
```

**Comment by mmcintyre123** [comment]
> t

function((?!^\}).|\n)+this is a string((?!^\}).|\n)+.

That expression works in Sublime Text.  Note - not all regex engines are the same.  You may have to tweak the expression depending on the context.  Also, there are many ways to skin a cat with regex.  Someone might be able to offer a better way. 

With the above, you can write any expression matching something from within the block that identifies it (in this case, "this is a string"), then surround that with the expression ((?!^\}).|\n)+, which says "match all lines until a line beginning with }".   That will then match the whole block.  The extra dot at the end includes the final } in the match.  You can create more layers of the sandwich if you need to quote from other lines to single out the blocks you want.
  
You don't need to master regex to the point that you can write it from scratch or in a perfectly optimized way every time to get use out of it.  I recommend creating a file with snippets like this which meet a need.  Over time you will use them again and again, and combine them in new ways to meet new needs.  The sites below are also very helpful.

https://www.regular-expressions.info/
https://regex101.com/

**Comment by mmcintyre123** [comment]
> > Sure no problem. So, when you say find a specific block of code, would you be searching for it exactly as written? Or would you be searching based on part of the contents? Please provide a sample block of code and I will write some regex for you.
> 
> Exactly as written, but even in the case of a partial search it could still be a block of code instead of a line or phrase.
> 
> Let's say for example the code below. I've intentionally screwed some spaces, since that can happen.
> 
> ```
> function whatever( argument ) {
> 
>   let myvar=null;
> 
>   if ( argument){
>     myvar = '';
>   }
>   else {
>     myvar='this is a string that is not unique to this section';
>   }
> 
>   return myvar;
> }
> ```

function((?!^\}).|\n)+this is a string((?!^\}).|\n)+.

That expression works in Sublime Text. Note - not all regex engines are the same. You may have to tweak the expression depending on the context. Also, there are many ways to skin a cat with regex. Someone might be able to offer a better way.

With the above, you can write any expression matching something from within the block that identifies it (in this case, "this is a string"), then surround that with the expression ((?!^}).|\n)+, which says "match all lines until a line beginning with }". That will then match the whole block. The extra dot at the end includes the final } in the match. You can create more layers of the sandwich if you need to quote from other lines to single out the blocks you want.

You don't need to master regex to the point that you can write it from scratch or in a perfectly optimized way every time to get use out of it. I recommend creating a file with snippets like this which meet a need. Over time you will use them again and again, and combine them in new ways to meet new needs. The sites below are also very helpful.

https://www.regular-expressions.info/
https://regex101.com/

**Comment by JacksonKearl** [comment]
@Louis7777 I recommend finding or creating an extension for this. The extension would contribute a command that takes the current selection, replaces all whitespace with a regex like `\W+` (some care may have to be taken to not introduce backtracking inefficiencies), then executes the command `workbench.action.findInFiles` with args `query` and `isRegex` set as appropriate.

**Comment by Louis7777** [comment]
@mmcintyre123 First of, honestly, many many thanks for putting the time and effort to reply and to craft that regex.

To be frank, I do have a reasonable amount of regex knowledge and I also use those couple of sites that you mentioned, but I'm not proficient enough to craft complex ones of my own (I usually modify and build on other people's solutions), as I haven't invested much time in it.

You are not the first one who suggested that I use regex for this - there's been a couple of other people too in another discussion. My main problem was not to craft the regex, although admittedly it would have been a problem, but to wrap my head around what exactly you are naturally suggesting to do with regex. To write a single regex that can cater for all my block searches? To write a very specific regex around the block of code where I'll have escaped everything that has to be escaped? Something else perhaps? So I guess I had to see one such expression in action to even understand what we're talking about.

Regarding the regex that you crafted, I will definitely save that one. But not to use it for my block searches, except perhaps for an emergency. If in another 5 years this functionality still isn't part of the core or no one else has made it into an extension, I'll invest time into learning how to make VSCode extensions, and I assume it will come in handy.

For the time being, I don't think a **specific regex** is a realistic solution, unless it is a **versatile regex** that can cater for all block searches and that can be used as a snippet. For instance, my search may not even begin with a "function", and I really just want to be able to paste in that block of code, rather than a specific element or string that identifies the block. Tweaking the regex depending on the context isn't realistic either. I mean, scratch all that, I'll just open another program, do the search and replace there, and then return back to VSCode. Which is a total shame, because VSCode has everything else that I personally need, and even more. It's just the search that is lacking some features.

@JacksonKearl There isn't any such extension, to my knowledge. As for creating it as an extension, it currently isn't time-feasible for me to learn how to create VCode extensions, but I will consider it if in a few years if no one else has made one (so I'll save your guidelines as they will come in handy). For the time being, I'll have to resort to other programs, complementary to my VSCode installation.

**Comment by mmcintyre123** [comment]
> @mmcintyre123 First of, honestly, many many thanks for putting the time and effort to reply and to craft that regex.
> 
> To be frank, I do have a reasonable amount of regex knowledge and I also use those couple of sites that you mentioned, but I'm not proficient enough to craft complex ones of my own (I usually modify and build on other people's solutions), as I haven't invested much time in it.
> 
> You are not the first one who suggested that I use regex for this - there's been a couple of other people too in another discussion. My main problem was not to craft the regex, although admittedly it would have been a problem, but to wrap my head around what exactly you are naturally suggesting to do with regex. To write a single regex that can cater for all my block searches? To write a very specific regex around the block of code where I'll have escaped everything that has to be escaped? Something else perhaps? So I guess I had to see one such expression in action to even understand what we're talking about.
> 
> Regarding the regex that you crafted, I will definitely save that one. But not to use it for my block searches, except perhaps for an emergency. If in another 5 years this functionality still isn't part of the core or no one else has made it into an extension, I'll invest time into learning how to make VSCode extensions, and I assume it will come in handy.
> 
> For the time being, I don't think a **specific regex** is a realistic solution, unless it is a **versatile regex** that can cater for all block searches and that can be used as a snippet. For instance, my search may not even begin with a "function", and I really just want to be able to paste in that block of code, rather than a specific element or string that identifies the block. Tweaking the regex depending on the context isn't realistic either. I mean, scratch all that, I'll just open another program, do the search and replace there, and then return back to VSCode. Which is a total shame, because VSCode has everything else that I personally need, and even more. It's just the search that is lacking some features.
> 
> @JacksonKearl There isn't any such extension, to my knowledge. As for creating it as an extension, it currently isn't time-feasible for me to learn how to create VCode extensions, but I will consider it if in a few years no one else has made one (so I'll save your guidelines as they will come in handy). For the time being, I'll have to resort to other programs, complementary to my VSCode installation.

I look at this type of thing not in terms of whether X nice feature exists, but in terms of what is the problem, what are all the possible approaches one can take to solve (or eliminate) it in a particular tool, and is there a reasonably efficient solution in the given tool.  For my purposes, regex in VSCode has been a reasonably easy and efficient solution.  It takes very little time to modify.  But that is definitely a personal preference thing.  That said, I notice just now that VSCode supports multi-line searches as well -- just highlight the block of code you're looking for and press CTRL+SHIFT+F.  It perfectly searches the project (or just open files if you click the new toggle) for that block.

**Comment by Louis7777** [comment]
> That said, I notice just now that VSCode supports multi-line searches as well -- just highlight the block of code you're looking for and press CTRL+SHIFT+F. It perfectly searches the project (or just open files if you click the new toggle) for that block. [Edit] although, I see what you mean in terms of varying whitespaces missing some blocks. In that case I would say, auto format the code before searching :)

I'm aware of that, I've been using it in certain occasions, but... add/remove a single space in there and it's over - won't find anything :(

**Comment by sickdyd** [comment]
@Louis7777 I just read a few of the latest posts, so apologize if my comment doesn't help: why not try using something like [Prettier](https://prettier.io/) to have consistent spacing / formatting across all files? This should sort out the problem of having more / less spaces in a specific block.

**Comment by Louis7777** [comment]
> 
> 
> @Louis7777 I just read a few of the latest posts, so apologize if my comment doesn't help: why not try using something like [Prettier](https://prettier.io/) to have consistent spacing / formatting across all files? This should sort out the problem of having more / less spaces in a specific block.

What if it isn't even me who wrote the software in which I am searching blocks of code? Also, what if it isn't even me who wrote the block of code that I am searching for? Maybe it was some random guy who pasted a fix on Stack Overflow or on some other site that has it's own code formatting and with whatever whitespace the guy liked or screwed up.

I believe it gets really tiring to having to format whatever I copy from somewhere or manually wrap it around a regex, unless there was a universal and versatile regex that I could have as a snippet to do it for me. Or, say, a toggle in VSCode's search field.

**Comment by JacksonKearl** [comment]
@Louis7777 I made an extension [here](https://marketplace.visualstudio.com/items?itemName=jakearl.search-ignore-whitespace), hope this helps.

**Comment by Louis7777** [comment]
> 
> 
> @Louis7777 I made an extension [here](https://marketplace.visualstudio.com/items?itemName=jakearl.search-ignore-whitespace), hope this helps.

Hey, it sure does! You're awesome man, thank you! It seems to work reasonably well too. I still wish that one day it becomes part of the core (hopefully in the form of a toggle), but till then this extension will do the job.

**Comment by ArturoDent** [comment]
I wrote an extension that deals with "some" of the issues raised above: [Find and Transform](https://marketplace.visualstudio.com/items?itemName=ArturoDent.find-and-transform&ssr=false#overview).

You can, in a setting, keybinding or context menu, scope the 'files to include' input to variables like `${file}, ${fileDirname}` and many more or the previous search's files with matches so you could perform successive searches on just the previous successful files.  All the other search args, like `'isRegex`, etc. are supported.  And various combinations of args can be saved as settings or keybindings.

**Comment by danielo515** [comment]
> # update for new version solution
> > 2020.06.09
> 
> ```shell
> Version: 1.56.2
> Commit: 054a9295330880ed74ceaedda236253b4f39a335
> Date: 2021-05-12T17:44:30.902Z
> Electron: 12.0.4
> Chrome: 89.0.4389.114
> Node.js: 14.16.0
> V8: 8.9.255.24-electron.0
> OS: Darwin x64 20.3.0
> ```
> 
> 1. search all files
> 
> ![image](https://user-images.githubusercontent.com/7291672/121217884-d8afcb00-c8b4-11eb-843a-c7bebf673fa1.png)
> 
> 1. only search open files
> 
> ![image](https://user-images.githubusercontent.com/7291672/121218072-009f2e80-c8b5-11eb-8c45-de774b0b99a5.png)
> 
> ## demo
> > 1. search all files ✅
> 
> ![image](https://user-images.githubusercontent.com/7291672/121220022-d6e70700-c8b6-11eb-9dc1-ca9139342396.png)
> 
> > 1. search open files ✅
> 
> ![image](https://user-images.githubusercontent.com/7291672/121219982-cf276280-c8b6-11eb-8e9f-b36da26d4ba7.png)

This does not work because VSCode seems to have a limit in the number of files it can open. And even if it were not the case, opening 500 files will have a huge performance impact

**Comment by danavangard** [comment]
The files to multi-search for code to be replaced *should already be open
via tabbed windows*. So when you multi-search and multi-replace code, the
edit reflects *only on already opened files*. I currently use TextWrangler
which flawlessly and swiftly works that way. C'mon VSCode, KISS !

**Comment by CreepyGnome** [comment]
So is the gist of this is that there is no way to search the scope of all open tabs especially ones that are new data and have never been saved so are not a real file in a known location yet?   

On a Mac I am trying to find a replacement for Notepad++ and one of the features I use with that all the time is searching all my open tabs, that they call on a button "Find All in All Opened Documents" VS Code has been fairly good as a replacement for my use as this appears to be the only feature missing is that I cannot search only the open tabs.

**Comment by JacksonKearl** [comment]
@CreepyGnome if I understand correctly we actually do support what you're asking for:
![image](https://user-images.githubusercontent.com/8586769/130502230-97ee7339-2377-4fec-ac36-4bf3b3d296ff.png)

**Comment by CreepyGnome** [comment]
Okay, that seems to work @JacksonKearl thanks alot!!

Feedback for VS Code UX team is that is a crappy UX, they shouldn't hide the information like that in a tool tip, and an OPEN BOOK icon is meaningless, couldn't even tell it was a toggle at first. I mean I tried clicking it many times before and nothing seemed to change other than I thought it is what showed the help text in the field it is part of as that is the only change visible when clicking it.

I mean a checkbox/toggle UNDER the field with a descriptive title would be much better and not easy to miss.

**Comment by JacksonKearl** [comment]
There are some ideas on how to make the state more visible at https://github.com/microsoft/vscode/issues/130412, I've placed that on the September milestone.

**Comment by danielo515** [comment]
Is this issue to ambitious? Will it make sense to open a separate issue for the "search on git modified files" for example?

**Comment by danavangard** [comment]
Rod, check out this link
<https://www.barebones.com/products/textwrangler/download.html> for
TextWrangler (earlier version) and BBedit (latest version) for Mac, both
from the same authors . Make sure your version of macOS works - read up on
details in site.

**Comment by yuval-gilboa** [comment]
@JacksonKearl 
The new search version is certainly an improvement, thanks for that. But the main point of this feature request is still missed. 

The original motivation behind this request was to address the inherent usability problem of having two different search mechanisms in vscode - one on the sidebar and one inside the editor - which results in two sets of search boxes, two sets of search result highlights (overlapping each other and confusing), two sets of keyboard navigation commands (F3 and F4), and a lot of redundant typing and clicking just to switch from one search mode to the other. Not to mention all the accessibility issues that this duality creates.

The new version solves only part of the problem (i.e., search in specific folder / open files), but can easily be extended to cover all the other aspects by changing the "Search only in Open Editors" button from a toggle button to a dropdown button that opens a menu like below. The "files to include" field becomes a combobox that allows quick switching between the different search scopes, including also the recently-used search paths. Up/down arrows can still be used to rotate through this list, but in addition it should be possible to assign keyboard shortcuts for the special search paths (current file, open files, etc).

If this is implemented then it becomes possible to do all searches from single central location, use same shortcut keys to navigate all kind of searches, and have single keystroke/click to switch between the search modes. 

5 years since this request was opened, i'm wandering whether this will ever come. 
Should a separate issue be opened for this?

![image](https://user-images.githubusercontent.com/4710322/152460752-71e2b898-5c60-4756-a891-e43afd1d771e.png)

**Comment by popod** [comment]
I completely agree with @yuval-gilboa. But I love being able to quickly search into the current opened file with a simple `ctrl+f` ! It's so useful that I don't think it should be removed.

**Comment by danavangard** [comment]
Good point from @yuval-gilboa. Also agree to @popod. The universally-known
concept of using ctrl+f must be made available at all times and applicable
to current viewed file. Perhaps a ctrl+F+O would make sense to also search
opened files? And maybe ctrl+F+F to search in the currently used folder?
Shortcuts like these would represent most used searching events, without
having to access a menu. And if one needs finer control, then access to a
menu would suffice, offering item-clicking and keyboard shortcut list of
all available search events.

**Comment by danavangard** [comment]
Good point @yuval-gilboa. Also agree to @popod. The universally-known
concept of using ctrl+f must be made available at all times and applicable
to current viewed file. Perhaps a ctrl+F+O would make sense to also search
opened files? And maybe ctrl+F+F to search in the currently used folder?
Shortcuts like these would represent most used searching events, without
having to access a menu. And if one needs finer control, then access to a
menu would suffice, offering item-clicking and keyboard shortcut list of
all available search events.

**Comment by cvchris** [comment]
This feature is very important for me

**Comment by yuval-gilboa** [comment]
Linking to #75220, which is basically the same request.

**Comment by jez9999** [comment]
Boy, would I love to have this feature.  My main use case is "search in git changes"; I quite frequently find myself searching for stuff in current git changes (whether staged or unstaged) but there isn't really a good way to do it right now.

> ![image](https://user-images.githubusercontent.com/4710322/152460752-71e2b898-5c60-4756-a891-e43afd1d771e.png)

I wish this mockup also included "current Git changes" for clarity!

**Comment by jerrygreen** [comment]
«modified files» makes enough sense to me

BTW, was just dreaming about this feature just now, and you bumped it 😄 my use case was just now: search through newly added `console.log` entries, not just opened files, not entire project, but modified files

but of course there's zillion other use-cases than just mine

**Comment by tjx666** [comment]
find a good extension to support in diff:

https://github.com/Darklaki/vsc-diff-regex

**Comment by mickae1** [comment]
> Boy, would I love to have this feature.  My main use case is "search in git changes"; I quite frequently find myself searching for stuff in current git changes (whether staged or unstaged) but there isn't really a good way to do it right now.
> 
> > ![image](https://user-images.githubusercontent.com/4710322/152460752-71e2b898-5c60-4756-a891-e43afd1d771e.png)
> 
> I wish this mockup also included "current Git changes" for clarity!

When I'm using the extension cmake I would love to search in all the files that the extension cmake Discovered.

**Comment by H-G-Hristov** [comment]
Add the ability to save (customize) searches and search locations. Particularly saving presets in "Files to include", "Files to exclude" would be extremely useful.
I often need to search in a very large project but I need to look only in a specific locations and I often need to switch between them. So I'd like to be able to predefine these searches and to quickly select them from a dropdown list when I click on "Files to include" or "Files to exclude".
Similar I'd like to have the ability to save presets in the search field to save any regexes.

I think this request deserves a separate github issue but I see that these get closed as duplicates so I'm typing it here.

**Comment by tjx666** [comment]
try my extension: 

https://github.com/tjx666/power-edit#search-in-git-changed-files

https://github.com/microsoft/vscode/assets/41773861/42c81901-da8c-4159-8de2-4342b36afad7

**Comment by zargold** [comment]
I came across this issue because I wanted to search for let's say all files that have PureComponent and within those files all instances of <Switch>. If that's your scenario no need for an extension or for this, but you do need to trick the regex search into giving you /m (multi-line search).


1. search for something simple like PureComponent (possibly targeting only files with the extension *.tsx, *.jsx)
2. Click the "Open in editor" link in blue under your search query.
---
This is because the next search is inefficient and it doesn't help to run it twice. (It can also confirm that the regex search is working as expected.)

3. Now while searching within editor (on the larger right hand panel): update your search query:
```
[\s\S\n]*PureComponent[\s\S\n]*
```

To find all files that use PureComponent (as an example) you can use the regex mode search: ([.*] button in the search)
- for some reason we need to redundantly target [\s\S\n]* before and after which gives you the full content of a file before and after (all characters or white-space including new lines before and after the exactly matching word "PureComponent").
4.  Results of this search will be vast since every character of each file will appear in results.
5. Now within the expanded results you can search for `<Switch>` using cmd+F search within Results view.

This option is pretty quick and easy to use, but your computer will struggle with how apparently inefficient the regex search is.

**Comment by Louis7777** [comment]
@zargold 

There's a popular plugin (~100K installs) for Visual Studio, but unfortunately nothing with GUI controls to ignore whitespace in Visual Studio Code.

The images below are from the "[Multiline Search and Replace](https://marketplace.visualstudio.com/items?itemName=PeterMacej.MultilineSearchandReplace)" plugin:

![search and replace](https://petermacej.gallerycdn.vsassets.io/extensions/petermacej/multilinesearchandreplace/3.5/1710759416233/126265/1/multiline_search_menu.png)

![ignore all whitespaces](https://petermacej.gallerycdn.vsassets.io/extensions/petermacej/multilinesearchandreplace/3.5/1710759416233/multiline_search_dialog__1.png)

The popularity of that plugin shows that there's a need. I wish ignoring whitespace was recognized as a basic feature by the team.
